No music (External, MIDI & MIDI Expansion)

### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Linux

### Describe the bug

When I start fheroes2 (.desktop or Terminal), it doesn't play music. However, effects are there.

### Save file

[Savefile.zip](https://github.com/ihhub/fheroes2/files/8466759/Savefile.zip)


### Additional info

Another user had the [problem](https://github.com/ihhub/fheroes2/issues/4755) before. However, this was on Windows. But I don't understand [oleg-derevenetz's solution](https://github.com/ihhub/fheroes2/issues/4755#issuecomment-997431886) (sorry). I could only take something with UTF. If it helps, I only use `de_DE.UTF-8 UTF-8` in my `/etc/locale.gen`. I also use Arch Linux with the latest updates. I installed fheroes2 from the [AUR](https://aur.archlinux.org/packages/fheroes2). I use `sdl2_mixer` from the Arch - Repository. But qark recommends `sdl2_mixer (sdl2_mixer-minimal-git)`. But when I did this, the game crashes after the intro. The console says that there are no write permissions. With `version 0.9.13` I also had music with `sdl2_mixer` from the Arch Repository before. I have also copied the required files / folders into `/home/marcel/.local/share/fheroes2/` again.

Here are the outputs of the console with the different Music Types:

Extern Output:

fheroes2
Free Heroes of Might and Magic II, version: 0.9.14
18:46:19: [ERROR]	Play:  Couldn't open '/home/marcel/.local/share/fheroes2/music/homm2_41.ogg'

MIDI Output:

fheroes2
Free Heroes of Might and Magic II, version: 0.9.14
18:50:49: [ERROR]	Play:  Couldn't open /etc/timidity++/freepats.cfg

MIDI Expansion Output:

fheroes2
Free Heroes of Might and Magic II, version: 0.9.14
18:52:33: [ERROR]	Play:  Couldn't open /etc/timidity++/freepats.cfg

The Intro plays sound on all three Music Types. As well as the effects. timidity++ is installed, but there is no timidity++ folder in `/etc/`. Just one folder called `timidity`, which contains a config file called `timidity.cfg`.

Hi @Monstanner

> Free Heroes of Might and Magic II, version: 0.9.14
> 18:46:19: [ERROR] Play: Couldn't open '/home/marcel/.local/share/fheroes2/music/homm2_41.ogg'

Do you really have this OGG file?

> Free Heroes of Might and Magic II, version: 0.9.14
> 18:50:49: [ERROR] Play: Couldn't open /etc/timidity++/freepats.cfg

It seems that SDL expects `timidity++` (software MIDI synthesizer) to play MIDI files, and it is absent on your machine (or improperly configured).

Hi @Monstanner

> Free Heroes of Might and Magic II, version: 0.9.14
> 18:46:19: [ERROR] Play: Couldn't open '/home/marcel/.local/share/fheroes2/music/homm2_41.ogg'

Do you really have this OGG file?

> Free Heroes of Might and Magic II, version: 0.9.14
> 18:50:49: [ERROR] Play: Couldn't open /etc/timidity++/freepats.cfg

It seems that SDL expects `timidity++` (software MIDI synthesizer) to play MIDI files, and it is absent on your machine (or improperly configured).

> Hi @Monstanner
> 
> > Free Heroes of Might and Magic II, version: 0.9.14
> > 18:46:19: [ERROR] Play: Couldn't open '/home/marcel/.local/share/fheroes2/music/homm2_41.ogg'
> 
> Do you really have this OGG file?

Yes I have. This is the output from `dir /home/marcel/.local/share/fheroes2/MUSIC/`:

homm2_01.ogg  homm2_10.ogg  homm2_19.ogg  homm2_28.ogg	homm2_37.ogg
homm2_02.ogg  homm2_11.ogg  homm2_20.ogg  homm2_29.ogg	homm2_38.ogg
homm2_03.ogg  homm2_12.ogg  homm2_21.ogg  homm2_30.ogg	homm2_39.ogg
homm2_04.ogg  homm2_13.ogg  homm2_22.ogg  homm2_31.ogg	homm2_40.ogg
homm2_05.ogg  homm2_14.ogg  homm2_23.ogg  homm2_32.ogg	homm2_41.ogg
homm2_06.ogg  homm2_15.ogg  homm2_24.ogg  homm2_33.ogg	homm2_42.ogg
homm2_07.ogg  homm2_16.ogg  homm2_25.ogg  homm2_34.ogg	pol
homm2_08.ogg  homm2_17.ogg  homm2_26.ogg  homm2_35.ogg	sw
homm2_09.ogg  homm2_18.ogg  homm2_27.ogg  homm2_36.ogg

This from `dir /home/marcel/.local/share/fheroes2/MUSIC/pol/`:

homm2_04.ogg  homm2_06.ogg  homm2_08.ogg
homm2_05.ogg  homm2_07.ogg  homm2_09.ogg

anf from `dir /home/marcel/.local/share/fheroes2/MUSIC/sw/`:

homm2_04.ogg  homm2_06.ogg  homm2_08.ogg
homm2_05.ogg  homm2_07.ogg  homm2_09.ogg

> 
> > Free Heroes of Might and Magic II, version: 0.9.14
> > 18:50:49: [ERROR] Play: Couldn't open /etc/timidity++/freepats.cfg
> 
> It seems that SDL Mixer expects `timidity++` (software MIDI synthesizer) to play MIDI files, and it is absent on your machine (or improperly configured).

Yes, I also find it strange because it is installed. No idea what it is, but in 0.9.13 the music was still there. I can't say if it was just at Extern. Because where I started fheroes2 for the first time, everything went normally. I left everything on standard, except for the resolution.

--- a/src/fheroes2/agg/agg.cpp
+++ b/src/fheroes2/agg/agg.cpp
@@ -83,7 +83,7 @@ namespace
             ListFiles musicFilePaths;
             musicFilePaths.ReadDir( dir, fileName, false );
             if ( musicFilePaths.empty() ) {
-               continue;
+                continue;
             }
 
             std::string correctFilePath = System::ConcatePath( dir, fileName );
@@ -610,7 +610,7 @@ void AGG::PlayMusicInternally( const int mus, const MusicSource musicType, const
 
         // To avoid extra I/O calls to data storage it might be useful to remember the last successful type of music and try to search for it next time.
         static std::array<MusicFileType, 3> musicFileTypes{ MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::DOS_VERSION ),
-            MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::WIN_VERSION ), MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::MAPPED ) };
+                                                            MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::WIN_VERSION ), MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::MAPPED ) };
 
         std::string filename;
 

--- a/src/fheroes2/agg/agg.cpp
+++ b/src/fheroes2/agg/agg.cpp
@@ -22,6 +22,7 @@
  ***************************************************************************/
 
 #include <algorithm>
+#include <array>
 #include <cassert>
 #include <condition_variable>
 #include <map>
@@ -48,6 +49,19 @@
 
 namespace
 {
+    struct MusicFileType
+    {
+        explicit MusicFileType( const MUS::EXTERNAL_MUSIC_TYPE type_ )
+            : type( type_ )
+        {
+            // Do nothing.
+        }
+
+        MUS::EXTERNAL_MUSIC_TYPE type = MUS::EXTERNAL_MUSIC_TYPE::WIN_VERSION;
+
+        std::array<std::string, 3> extension{ ".ogg", ".mp3", ".flac" };
+    };
+
     const std::string externalMusicDirectory( "music" );
 
     std::vector<std::string> getMusicDirectories()
@@ -63,44 +77,56 @@ namespace
         return directories;
     }
 
-    std::string getExternalMusicFile( const int musicTrackId, const MUS::EXTERNAL_MUSIC_TYPE musicType, const std::vector<std::string> & directories )
+    bool findMusicFile( const std::vector<std::string> & directories, const std::string & fileName, std::string & fullPath )
+    {
+        for ( const std::string & dir : directories ) {
+            ListFiles musicFilePaths;
+            musicFilePaths.ReadDir( dir, fileName, false );
+            if ( musicFilePaths.empty() ) {
+               continue;
+            }
+
+            std::string correctFilePath = System::ConcatePath( dir, fileName );
+            correctFilePath = StringLower( correctFilePath );
+
+            for ( const std::string & path : musicFilePaths ) {
+                const std::string temp = StringLower( path );
+                if ( temp == correctFilePath ) {
+                    fullPath = std::move( correctFilePath );
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    std::string getExternalMusicFile( const int musicTrackId, const std::vector<std::string> & directories, MusicFileType & musicType )
     {
         if ( directories.empty() ) {
             // Nothing to search.
             return {};
         }
 
-        // Instead of relying some generic functions we want to have maximum performance as I/O operations are the slowest.
-        std::vector<std::string> possibleFilenames;
-        possibleFilenames.reserve( directories.size() );
+        std::string fullPath;
 
-        for ( const std::string & dir : directories ) {
-            possibleFilenames.emplace_back( System::ConcatePath( dir, MUS::getFileName( musicTrackId, musicType, ".flac" ) ) );
+        std::string fileName = MUS::getFileName( musicTrackId, musicType.type, musicType.extension[0].c_str() );
+        if ( findMusicFile( directories, fileName, fullPath ) ) {
+            return fullPath;
         }
 
-        // Search for FLAC files as they have the best audio quality.
-        for ( const std::string & filename : possibleFilenames ) {
-            if ( System::IsFile( filename ) ) {
-                return filename;
-            }
+        fheroes2::replaceStringEnding( fileName, musicType.extension[0].c_str(), musicType.extension[1].c_str() );
+        if ( findMusicFile( directories, fileName, fullPath ) ) {
+            // Swap extensions to improve cache hit.
+            std::swap( musicType.extension[0], musicType.extension[1] );
+            return fullPath;
         }
 
-        // None of FLAC files found. Try OGG files.
-        for ( std::string & filename : possibleFilenames ) {
-            fheroes2::replaceStringEnding( filename, ".flac", ".ogg" );
-
-            if ( System::IsFile( filename ) ) {
-                return filename;
-            }
-        }
-
-        // No luck with even OGG. Try with MP3.
-        for ( std::string & filename : possibleFilenames ) {
-            fheroes2::replaceStringEnding( filename, ".ogg", ".mp3" );
-
-            if ( System::IsFile( filename ) ) {
-                return filename;
-            }
+        fheroes2::replaceStringEnding( fileName, musicType.extension[1].c_str(), musicType.extension[2].c_str() );
+        if ( findMusicFile( directories, fileName, fullPath ) ) {
+            // Swap extensions to improve cache hit.
+            std::swap( musicType.extension[0], musicType.extension[2] );
+            return fullPath;
         }
 
         // Looks like music file does not exist.
@@ -582,13 +608,21 @@ void AGG::PlayMusicInternally( const int mus, const MusicSource musicType, const
     if ( musicType == MUSIC_EXTERNAL ) {
         const std::vector<std::string> & musicDirectories = getMusicDirectories();
 
-        std::string filename = getExternalMusicFile( mus, MUS::EXTERNAL_MUSIC_TYPE::DOS_VERSION, musicDirectories );
-        if ( filename.empty() ) {
-            filename = getExternalMusicFile( mus, MUS::EXTERNAL_MUSIC_TYPE::WIN_VERSION, musicDirectories );
-        }
+        // To avoid extra I/O calls to data storage it might be useful to remember the last successful type of music and try to search for it next time.
+        static std::array<MusicFileType, 3> musicFileTypes{ MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::DOS_VERSION ),
+            MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::WIN_VERSION ), MusicFileType( MUS::EXTERNAL_MUSIC_TYPE::MAPPED ) };
 
-        if ( filename.empty() ) {
-            filename = getExternalMusicFile( mus, MUS::EXTERNAL_MUSIC_TYPE::MAPPED, musicDirectories );
+        std::string filename;
+
+        for ( size_t i = 0; i < musicFileTypes.size(); ++i ) {
+            filename = getExternalMusicFile( mus, musicDirectories, musicFileTypes[i] );
+            if ( !filename.empty() ) {
+                if ( i > 0 ) {
+                    // Swap music types to improve cache hit.
+                    std::swap( musicFileTypes[0], musicFileTypes[i] );
+                }
+                break;
+            }
         }
 
         if ( filename.empty() ) {

--- a/src/fheroes2/agg/agg.cpp
+++ b/src/fheroes2/agg/agg.cpp
@@ -34,18 +34,15 @@
 #include "agg_file.h"
 #include "audio.h"
 #include "dir.h"
-#include "embedded_image.h"
 #include "game.h"
 #include "localevent.h"
 #include "logging.h"
 #include "m82.h"
 #include "mus.h"
-#include "screen.h"
 #include "settings.h"
 #include "system.h"
 #include "tools.h"
 #include "xmi.h"
-#include "zzlib.h"
 
 namespace
 {

--- a/src/fheroes2/agg/agg.cpp
+++ b/src/fheroes2/agg/agg.cpp
@@ -89,7 +89,7 @@ namespace
             for ( const std::string & path : musicFilePaths ) {
                 const std::string temp = StringLower( path );
                 if ( temp == correctFilePath ) {
-                    fullPath = std::move( correctFilePath );
+                    fullPath = path;
                     return true;
                 }
             }

